### PR TITLE
feat(sentry): use WeakMap cache for printed documents

### PR DESCRIPTION
**Pull request by jeengbe**
🚨 **IMPORTANT: Please do not create a Pull Request without creating an issue first.**   welll  ¯\_(ツ)_/¯ why would you discourage MRs

_Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request._

## Description

In scenarios where the used document object is cached, and therefore the same by reference, an option is provided to opt in to an LRU cache so that the document doesn't need to be reparsed over and over.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Screenshots/Sandbox (if appropriate/relevant):

n/a

## How Has This Been Tested?

Unit tests should pass

**Test Environment**:

n/a

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: packages/plugins/sentry/src/index.ts [hunk 1]
```diff
@@ -92,6 +104,12 @@ export const useSentry = (options: SentryPluginOptions = {}): Plugin => {
   const renameTransaction = pick('renameTransaction', false);
   const skipOperation = pick('skip', () => false);
   const skipError = pick('skipError', defaultSkipError);
+  const cachePrintedDocuments = pick('cachePrintedDocuments', false);
```
Comment (ardatan) [inline packages/plugins/sentry/src/index.ts]: I think caching the document from `onParse` might be better.

Comment (jeengbe) [reply packages/plugins/sentry/src/index.ts]: What do you mean? This MR is about caching the "expensive" `print()`, not any parsing.

Comment (ardatan) [reply packages/plugins/sentry/src/index.ts]:
```ts
const documentSDLMap = new WeakMap<DocumentNode, string>();
{
  onParse({ params }) {
      return ({ result }) => {
         documentSDLMap.set(result, params.source.toString());
      }
  },
  onExecute({ args }) {
    // Just in case `parse` is not called by the server, fallback to `print`
    const sdl = documentSDLMap.get(args.document) || print(sdl);
  }
}
```

Comment (jeengbe) [reply packages/plugins/sentry/src/index.ts]: The idea is good, but sadly this is not quite the same, as the result from `print()` is formatted whereas this could be any raw minified input sdl.

Comment (jeengbe) [reply packages/plugins/sentry/src/index.ts]: Also, not exactly what I was aiming for in this MR, but thinking about it, a WeakMap might be the right choice after all. I initially decided against it due to its uselessness with freshly parsed documents. 

My use case for this MR is that with a persisted operations plugin, the documents object is always the same by reference for most queries, which is a golden opportunity for caching.

Comment (ardatan) [reply packages/plugins/sentry/src/index.ts]: Persisted operations plugin also set the `DocumentNode` so the same object is used then it works exactly like you said if I don't miss anything.

Comment (jeengbe) [reply packages/plugins/sentry/src/index.ts]: Yeah, but my point is that you can't use `params.source.toString()` since that's most likely a non-indented blob input, so there needs to be at least one call to `print()`, which might as well be this one.

**Review by ardatan — APPROVED** [review]

File: packages/plugins/sentry/src/index.ts [hunk 2]
```diff
@@ -157,7 +165,7 @@ export const useSentry = (options: SentryPluginOptions = {}): Plugin => {
           // eslint-disable-next-line no-console
           console.warn(
             [
-              `Flag "startTransaction" is enabled but Sentry failed to find a transaction.`,
+              `Flag "startTransaction" is disabled but Sentry failed to find a transaction.`,
```
Comment (ardatan) [inline packages/plugins/sentry/src/index.ts:168]: I think this is not related :)

**Review by ardatan — CHANGES_REQUESTED** [review]

File: packages/plugins/sentry/src/index.ts (hunk 2 above)
Comment (jeengbe) [reply packages/plugins/sentry/src/index.ts:168]: It's not, I just noticed it and thought I'd fix it while I'm at it :]

**Review by n1ru4l — APPROVED** [review]
